### ISSUE TITLE
`torch.linalg.solve` on GPU gives "Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device"

### ISSUE THREAD

**Issue by ilija139** [pytorch/pytorch#60175]
## 🐛 Bug

`torch.linalg.solve` on GPU gives "Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device"
But it still gives correct results. 
I tried on two machines, same result. 
## To Reproduce

```
Python 3.8.10 (default, Jun  4 2021, 15:09:15) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.9.0a0+gitd69c22d'
>>> A = torch.randn(2,3,3).cuda()
>>> B = torch.randn(2,3,4).cuda()
>>> X = torch.linalg.solve(A, B)
Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device
Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device
```

## Expected behavior
No error to be printed. `X` is already correctly computed. 

### Machine 1
Collecting environment information...
PyTorch version: 1.9.0a0+gitd69c22d
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Pop!_OS 20.10 (x86_64)
GCC version: (Ubuntu 10.2.0-13ubuntu1) 10.2.0
Clang version: Could not collect
CMake version: version 3.20.4
Libc version: glibc-2.32

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.11.0-7614-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.0a0+gitd69c22d
[conda] blas                      1.0                         mkl  
[conda] magma-cuda112             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-include               2021.2.0           h06a4308_296  
[conda] mkl-service               2.4.0            py38h497a2fe_0    conda-forge
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.20.2           py38h2d18471_0  
[conda] numpy-base                1.20.2           py38hfae3a4d_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.9.0a0+gitd69c22d          pypi_0    pypi

### Machine 2
Collecting environment information...
PyTorch version: 1.9.0a0+gitd69c22d
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Pop!_OS 20.10 (x86_64)
GCC version: (Ubuntu 10.2.0-13ubuntu1) 10.2.0
Clang version: Could not collect
CMake version: version 3.20.4
Libc version: glibc-2.32

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.11.0-7614-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: Quadro RTX 4000
Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] gpytorch==1.4.2
[pip3] numpy==1.20.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.0a0+gitd69c22d
[conda] blas                      1.0                         mkl  
[conda] gpytorch                  1.4.2                    pypi_0    pypi
[conda] magma-cuda112             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-include               2021.2.0           h06a4308_296  
[conda] mkl-service               2.4.0            py38h497a2fe_0    conda-forge
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.20.2           py38h2d18471_0  
[conda] numpy-base                1.20.2           py38hfae3a4d_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.9.0a0+gitd69c22d          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @malfet @seemethere @walterddr @ngimel @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @IvanYashchuk @xwang233 @Lezcano

**Comment by zou3519** [comment]
Tentatively marking this as hi-pri for not being able to use torch.linalg.solve on GPU, but since we have tests for these this might be an environment issue

**Comment by ilija139** [comment]
@zou3519 Note that it gives a correct solution, so it works and the test will pass, but it will still print the error message. 
I installed pytorch from source using the tag v1.9.0 version.

**Comment by ilija139** [comment]
It works fine when installed via `conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c nvidia`

Even though my CUDA version is 11.2 (hence the reason why I compiled from source).
The env of the *working* version:
```
Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Pop!_OS 20.10 (x86_64)
GCC version: (Ubuntu 10.2.0-13ubuntu1) 10.2.0
Clang version: Could not collect
CMake version: version 3.20.4
Libc version: glibc-2.32

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.11.0-7614-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: Quadro RTX 4000
Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] gpytorch==1.4.2
[pip3] numpy==1.20.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.0
[pip3] torchaudio==0.9.0a0+33b2469
[pip3] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] gpytorch                  1.4.2                    pypi_0    pypi
[conda] magma-cuda112             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h06a4308_296
[conda] mkl-include               2021.2.0           h06a4308_296
[conda] mkl-service               2.4.0            py38h497a2fe_0    conda-forge
[conda] mkl_fft                   1.3.0            py38h42c9631_2
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.20.2           py38h2d18471_0
[conda] numpy-base                1.20.2           py38hfae3a4d_0
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1
[conda] pytorch                   1.9.0           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchaudio                0.9.0                      py38    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch
```

```
▶ nvidia-smi
Fri Jun 18 20:03:56 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.73.01    Driver Version: 460.73.01    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro RTX 4000     On   | 00000000:01:00.0  On |                  N/A |
| 30%   42C    P8    12W / 125W |      1MiB /  7948MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

```

It also prints `arch < 200 not supported` when I use larger matrices, e.g.:
```
X = torch.linalg.solve(torch.randn(200,200,200).cuda(), torch.randn(200,200,200).cuda())
Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device
arch < 200 not supported 
```

**Comment by malfet** [comment]
@ilija139 can you please elaborate how you've build the magma?

**Comment by ilija139** [comment]
@malfet  `conda install -c pytorch magma-cuda112`

**Comment by IvanYashchuk** [comment]
Internally MAGMA in various places MAGMA has code like:
```cpp
magma_int_t arch = magma_getdevice_arch();
if (arch condition) {
} else {
}
```
`magma_getdevice_arch` returns `CUDA_ARCH` if `magma_init();` was called before (`magma_init` caches some information about the system) or returns 0 and prints to stderr a message reported above. Since it returns 0 and there's a fallback "else" branch it's harmless.

To disable the message `magma_init` should be called in PyTorch. `magma_init` is called together with CUDA initialization
https://github.com/pytorch/pytorch/blob/407c6372a8ef0a8b63f14b04f688e4cad39496fd/aten/src/ATen/cuda/detail/CUDAHooks.cpp#L55-L57
@malfet, maybe `aten/src/THC/THCTensorMathMagma.cu` is skipped in compilation leaving `at::cuda::detail::THCMagma_init` uninitialized?
https://github.com/pytorch/pytorch/blob/407c6372a8ef0a8b63f14b04f688e4cad39496fd/aten/src/THC/THCTensorMathMagma.cu#L28-L32

**Comment by mruberry** [comment]
cc @ngimel, too, who's removing significant parts of TH*.

**Comment by IvanYashchuk** [comment]
I can reproduce the problem in `v1.9.0` tag of PyTorch built from source:
```python
>>> import torch
>>> torch.__version__
'1.9.0a0+gitd69c22d'
>>> A = torch.randn(2,3,3).cuda()
>>> B = torch.randn(2,3,4).cuda()
>>> X = torch.linalg.solve(A, B)
Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device
Error in magma_getdevice_arch: MAGMA not initialized (call magma_init() first) or bad device
```
with magma from pytorch conda channel. However, the message is not present with the latest master branch, so I believe the problem is fixed.

**Comment by IvanYashchuk** [comment]
If someone is building from source 1.9.0 and would like to suppress the warning, cherry-pick [this commit](https://github.com/pytorch/pytorch/commit/296d2a4399923370dccc5f3f4fc0f5fd12db4d19). It should fix the magma initialization and remove the "Error in magma_getdevice_arch:" message.

**Comment by johan12345** [comment]
I am now running into this bug with the [PyPI version of PyTorch 1.9.1](https://pypi.org/project/torch/#files) for Linux (released yesterday).  It does not occur with the 1.9.0 version.

**Comment by IvanYashchuk** [comment]
Unfortunately, 1.9.1 doesn't include [the commit](https://github.com/pytorch/pytorch/commit/296d2a4399923370dccc5f3f4fc0f5fd12db4d19) that fixed the issue. The next release that will have the problem fixed is 1.10.0

**Comment by johan12345** [comment]
Ah, thanks! Still strange that it worked in 1.9.0 from PyPI though...
